doc/format-run.md

```diff
@@ -0,0 +1,216 @@
+  distributed). Initially we will support a compact index type suited to
+  uniformly distributed keys of variable size but with at least 48bits.
+* The compact index "range finder" bits (if a compact index is used).
+
+## Key/operations file
+
+This file contains all the key/operation/blobref entries for the LSM run,
+sorted by key and arranged into pages. The individual page format is detailed
+in `format-page.md`.
+
+The overall file format consists exclusivly of these pages, with no additional
+metadata or file headers. The page size will typically be 4k but can be any
+power of 2 up to 64k.
+
+The run is sorted by key, based on the usual key byte string representation.
+The number of each disk page within the file is significant, as this page
+number is what is stored within the LSM run index, to enable easily finding the
+right disk page(s) to read for a lookup operation. Pages are numbered from 0.
+
+The key/operations are organised into pages in the following way:
+
+* All the keys in a page must have the same "range finder" bits, if a compact
+  index is in use. For code uniformity, if a compact index is not used then
+  this is equivalent to using 0 range finder bits. This restriction is needed
+  to support the compact index representation.
+  
+  When packing key/operations into pages linearly, the initial key bits can be
+  tracked and when they change from one key to the next, then a page boundary
+  must be used, at the expense of wasting any remaining space in the page. The
+  range finder bits are chosen to not cause excessive waste.
+
+* Otherwise, key/operation/blobrefs are packed into pages so that as many fit
+  as possible, given the size limit of the page format. The details of the size
+  of pages is given in `format-page.md`.
+
+* For the special case of N=1, the value can be large and span multiple pages.
+
+* In all other cases, N>1, all entries must fit within a single page.
+
+### Maximum key and value sizes
+
+The maximum key size is limited by what can fit within a page, thus the
+maximum key size depends on the page size. The page size overhead (derived in
+`format-page.md`) is 44 bytes. We can however round to more "sensible" looking
+numbers.
+
+* 4k pages:  2^12-44, rounded: 4,000
+* 8k pages,  2^13-44, rounded: 8,000
+* 16k pages, 2^14-44, rounded: 16,000
+* 32k pages, 2^15-44, rounded: 32,000
+* 64k pages, 2^16-44, rounded: 64,000
+
+The maximum value size is 2^32-1 less the page overhead, and conservatively
+assuming a maximum sized key with maximum page size. This is approximately
+2^32 - 2^16 - 1 = 4294901759, though this limit could also be rounded down in
+a public API.
+
+## Blobs file
+
+Each key/operation in the key/operations file optionally contains a blob
+references. A blob reference is of course a reference to a blob value, which
+is simply a sequence of bytes. These blob values are stored in the blobs file.
+
+The representation of a blob reference is a 64bit byte file offset and a 32bit
```

Comment:
I initially interpreted this as "a 64-bit byte", instead of a file offset in bytes, with all possible values represented in 64 bits :P

Reply:
clarified.